Kernel/Net/NetworkAdapter.cpp

```diff
@@ -153,27 +153,34 @@ void NetworkAdapter::did_receive(ReadonlyBytes payload)
     m_packets_in++;
     m_bytes_in += payload.size();
 
-    Optional<KBuffer> buffer;
-
     if (m_packet_queue_size == max_packet_buffers) {
         // FIXME: Keep track of the number of dropped packets
         return;
     }
 
-    if (m_unused_packet_buffers.is_empty()) {
-        buffer = KBuffer::copy(payload.data(), payload.size());
+    RefPtr<PacketWithTimestamp> packet;
+
+    if (m_unused_packets.is_empty()) {
+        auto buffer = KBuffer::copy(payload.data(), payload.size());
+        packet = adopt_ref_if_nonnull(new PacketWithTimestamp { move(buffer), kgettimeofday() });
     } else {
-        buffer = m_unused_packet_buffers.take_first();
-        --m_unused_packet_buffers_count;
-        if (payload.size() <= buffer.value().capacity()) {
-            memcpy(buffer.value().data(), payload.data(), payload.size());
-            buffer.value().set_size(payload.size());
+        packet = m_unused_packets.take_first();
+        if (payload.size() <= packet->buffer.capacity()) {
+            memcpy(packet->buffer.data(), payload.data(), payload.size());
```

Comment:
Is there ever a case where the payload and the packet buffer would be different sizes?